lib/Frontend/CachingUtils.cpp

```diff
@@ -262,53 +253,41 @@ bool replayCachedCompilerOutputs(
     // Add symbol graph entry for lookup. Output path doesn't matter here.
     Outputs.try_emplace(file_types::ID::TY_SymbolGraphFile, "<symbol-graph>");
 
-    return replayOutputsForInputFile(Input, InputPath, InputIndex, Outputs);
+    return replayOutputsForInputFile(Input, InputPath, InputIndex, OutputRef,
+                                     Outputs);
   };
 
-  auto AllInputs = Opts.InputsAndOutputs.getAllInputs();
-  // If there are primary inputs, look up only the primary input files.
-  // Otherwise, prepare to do cache lookup for all inputs.
-  for (unsigned Index = 0; Index < AllInputs.size(); ++Index) {
-    const auto &Input = AllInputs[Index];
-    if (Opts.InputsAndOutputs.hasPrimaryInputs() && !Input.isPrimary())
-      continue;
-
-    replayOutputFromInput(Input, Index);
-  }
+  for (auto In : Inputs)
+    replayOutputFromInput(In.Input, In.Index, In.OutputRef);
 
   if (!CanReplayAllOutput)
     return false;
 
-  // If there is not diagnostic output, this is a job that produces no output
-  // and only diagnostics, like `typecheck-module-from-interface`, look up
-  // diagnostics from first file.
-  if (!DiagnosticsOutput)
-    replayOutputsForInputFile(
-        Opts.InputsAndOutputs.getFirstOutputProducingInput(),
-        "<cached-diagnostics>",
-        Opts.InputsAndOutputs.getIndexOfFirstOutputProducingInput(),
-        {{file_types::ID::TY_CachedDiagnostics, "<cached-diagnostics>"}});
-
-  // Check again to make sure diagnostics is fetched successfully.
-  if (!CanReplayAllOutput)
+  auto failedReplay = [DiagHelper]() {
+    if (DiagHelper)
+      DiagHelper->endMessage(/*retCode=*/1);
```

Comment:
The equivalent code from `replayCompilation` in `SwiftCaching.cpp` called `Inst.getDiags().finishProcessing();` here and also below in the successful case.  Do we need something similar here?

Reply:
`finishProcessing()` is moved to call site after calling the replay function, with or without diagnostics emitted, since it doesn't hurt to call it even there are no diagnostics emitted.